catroid/src/main/java/org/catrobat/catroid/content/actions/SpeakAction.kt

```diff
@@ -0,0 +1,117 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.content.actions
+
+import android.speech.tts.TextToSpeech
+import android.speech.tts.UtteranceProgressListener
+import android.util.Log
+
+import com.badlogic.gdx.scenes.scene2d.actions.TemporalAction
+
+import org.catrobat.catroid.common.Constants
+import org.catrobat.catroid.content.Sprite
+import org.catrobat.catroid.formulaeditor.Formula
+import org.catrobat.catroid.formulaeditor.FormulaElement
+import org.catrobat.catroid.formulaeditor.InterpretationException
+import org.catrobat.catroid.io.SoundManager
+import org.catrobat.catroid.stage.TextToSpeechHolder
+import org.catrobat.catroid.utils.Utils
+
+import java.io.File
+import java.util.HashMap
+
+class SpeakAction : TemporalAction() {
+    private var text: Formula? = null
+    private var interpretedText: Any? = null
+    private var hashText: String? = null
+    private var sprite: Sprite? = null
+    lateinit var speechFile: File
+        private set
+    private lateinit var listener: UtteranceProgressListener
+    private var determineLength = false
+    var lengthOfText = 0f
+        private set
+
+    override fun begin() {
+        interpretFormula()
+        hashText = Utils.md5Checksum(interpretedText.toString())
+        val fileName = hashText
+        val pathToSpeechFile = File(Constants.TEXT_TO_SPEECH_TMP_PATH)
+        pathToSpeechFile.mkdirs()
+        speechFile = File(pathToSpeechFile, fileName + Constants.DEFAULT_SOUND_EXTENSION)
+        listener = object : UtteranceProgressListener() {
+            override fun onDone(utteranceId: String) {
+                if (determineLength) {
+                    lengthOfText = SoundManager.getInstance().getDurationOfSoundFile(speechFile.absolutePath)
+                } else {
+                    SoundManager.getInstance().playSoundFile(speechFile.absolutePath, sprite)
+                }
+            }
+            override fun onStart(utteranceId: String) {
+                Log.d(javaClass.simpleName, "onStart()")
+            }
+            override fun onError(utteranceId: String) {
+                Log.d(javaClass.simpleName, "onError()")
+            }
+        }
+        super.begin()
+    }
+
+    override fun update(delta: Float) {
+        val speakParameter = HashMap<String?, String?>()
+        speakParameter[TextToSpeech.Engine.KEY_PARAM_UTTERANCE_ID] = hashText
+        TextToSpeechHolder.instance?.textToSpeech(interpretedText.toString(), speechFile, listener, speakParameter)
+    }
+
+    fun setSprite(sprite: Sprite?) {
```

Comment:
Is this setter really useful?
Please make sprite public.
In Kotlin Getter and setter are automatically generated.
https://kotlinlang.org/docs/reference/properties.html#properties-and-fields